File: gomoku_cli.py
"""
Gomoku Player
"""
# import numpy as np
import math
from random import shuffle
from copy import deepcopy

X = 1 # White Player
O = -1 # Black Player
EMPTY = 0

def initial_state(size=15):
    """
    Returns starting state of the board.
    """
    # return np.array([[EMPTY]*size]*size)
    return [[EMPTY for _ in range(size)] for _ in range(size)]


def player_turn(board, last_player=None):
    """
    Returns player who has the next turn on a board.
    """
    if last_player == X or last_player == O:
        return X if last_player == O else O
    
    # return (O if np.sign(np.sum(board)) == X else X) if (board==EMPTY).any() else EMPTY

    _sum = 0 # sum of all elements in the board
    for row in board:
        _sum += sum(row)
    _sum_sign = int(math.copysign(1, _sum))

    any_empty = False
    for row in board:
        if EMPTY in row:
            any_empty = True
            break
    
    return (O if _sum_sign == X else X) if any_empty else EMPTY

def actions(board):
    """
    Returns set of all possible actions (i, j) on the board.
    """
    # return list(zip(*np.where(board == EMPTY)))
    board_len = len(board[0])
    return [(i, j) for i in range(board_len) for j in range(board_len) if board[i][j] == EMPTY]

def make_move(board, action, player=None):
    """
    Returns board after the move.
    """
    if player is None:
        player = player_turn(board)
    i, j = action
    # board[i, j] = player
    board[i][j] = player
    return board

def result(board, action, player=None):
    """
    Returns the board that results from making move (i, j) on the board.
    """
    if player is None:
        player = player_turn(board)
    return make_move(deepcopy(board), action, player)

def get(board, action):
    """
    Returns the value of board at action.
    """
    # board_len = board.shape[0]
    board_len = len(board[0])
    row, col = action
    
    if row < 0 or row >= board_len or col < 0 or col >= board_len: return 0
    # return board[row, col]
    return board[row][col]

def winner(board):
    """
    Returns the winner of the game, if there is one.
    """
    dirs = ((1, -1), (1, 0), (1, 1), (0, 1))
    # board_len = board.shape[0]
    board_len = len(board[0])
    
    for i in range(board_len):
        for j in range(board_len):
            if board[i][j] == EMPTY: continue
            id = board[i][j]
            for d in dirs:
                x, y = i, j
                count = 0
                while get(board, (x, y)) == id:
                    if count > 5 or get(board, (x, y)) != id: break
                    y += d[0]
                    x += d[1]
                    count += 1
                # Exactly 5 in a row and no more than 5
                if count == 5 and get(board, (i+5*d[0], j+5*d[1])) != id and get(board, (i-d[0], j-d[1])) != id:
                    return id
    return EMPTY

def terminal(board):
    """
    Returns True if game is over, False otherwise.
    """
    # return winner(board) != EMPTY or not (board == EMPTY).any()
    any_empty = False
    for row in board:
        if EMPTY in row:
            any_empty = True
            break

    return winner(board) != EMPTY or not any_empty

def display_state(board):
    board_len = len(board[0])
    
    print(' '+' -'*board_len)
    
    print('  ' + ' '.join([chr(ord('A') + i) for i in range(board_len)]), end="__\n")

    for row in range(board_len):
        print(chr(ord('A') + row), end=' ')
        for col in range(board_len):
            ch = board[row][col]
            if ch == EMPTY: 
                print('.', end=' ')
            elif ch == X:
                print( 'X', end=' ')
            elif ch == O:
                print( 'O', end=' ')
        print('|')
    if terminal(board):
        victor = winner(board)
        if victor != EMPTY:
            msg='|Player {} wins.'.format('O' if victor == X else 'X')
        else:
            msg='|draw'
    else:
        turn = 'X' if player_turn(board) == X else 'O'
        msg='|Player {}\'s turn'.format(turn)
    print(msg.ljust(31), end=' ')
    print("|_____")
    print("|___________________________________ |")
    print("                                    ↓↓")

def ai_move(board):
    """
    # TODO
    makes minimax converge faster by:
    1: make a 6 x 6 filter kernel and apply that filter to game board and perform these filter operations,
    2: rotate the filter(by 90°) 4 times and store those 4 filter configurations.
    3: convolve through the board parallelly(not sequentially) using all 4 filters.
    4: if any of the 4 filter matches the board subset during convolution, then
    5: apply minimax to find winning move(location) on the filtered subset of the board
    6: rotate the move/location according to filter configuration."
    7: calculate and assign (TODO: some kind of score) to the move to find the optimal(max/min) scoring move from the (board_size x board_size)moves obtained after entire convolvolution.
    8: return the optimal(minimum/maximum) move
    """
    pass

def minimax(board):
    """
    Returns the optimal action for the current player on the board.
    """
    board_len = len(board[0])
    if terminal(board):
        return None
    # find which pllayer's turn it is in the board
    turn = player_turn(board) == X
    # find the playable moves init dict to map value to optimised moves
    moves, v2action = actions(board), {}
    # shuffle for randomness
    shuffle(moves)

    minimum_score, maximum_score = -(board_len**2), board_len**2
    
    def alphabeta(b, alpha, beta, maximizingPlayer):
        """ min_/max_value function with alpha-beta pruning"""

        if terminal(b):
            return winner(b)

        else:
            score, optimizer = (minimum_score, max) if maximizingPlayer else (maximum_score, min)

            possible_moves = actions(b)

            for move in possible_moves:
                child = result(board=b, action=move)

                score = optimizer(score, alphabeta(b=child, alpha=alpha, beta=beta, maximizingPlayer=not maximizingPlayer))

                if maximizingPlayer:
                    alpha = optimizer(alpha, score)
                else:
                    beta = optimizer(beta, score)

                if alpha >= beta:
                    break

            return score
    
    # picks action a in ACTIONS(s) that produces optimal value of MIN/MAX-VALUE(RESULT(s, a))
    for move in moves:
        possible_board = result(board=board, action=move)
        v2action[alphabeta(b=possible_board, maximizingPlayer=not turn, alpha=minimum_score, beta=maximum_score)] = move
    # choose our optimizer function (min or max) based on which player is playing
    optimizer = max if turn else min
    # return the optimised action from dict.
    return v2action[optimizer(v2action)]


#----------------------------------------------------------------------
# main game
#----------------------------------------------------------------------
def gamemain():
    board_size = 15
    board = initial_state(board_size)

    while 1:
        row = col = -1
        display_state(board)
        turn = player_turn(board)
        turn_chr = 'X' if turn == X else 'O'
        action = input('Enter your move (q/f:Quit/Forfeit): ').strip('\r\n\t ').lower()
        
        if action == 'f':
            print('player {} forfeits.'.format(turn_chr))
            print('player {} wins.'.format('O' if turn == X else 'X'))
            
            print("new game")
            board = initial_state(board_size)
            continue
        
        elif action == 'q':
            print('player {} quits.'.format(turn_chr))
            break

        elif len(action) == 2:
            tr = ord(action[0].upper()) - ord('A')
            tc = ord(action[1].upper()) - ord('A')
            if tr >= 0 and tc >= 0 and tr < board_size and tc < board_size:
                if board[tr][tc] == 0:
                    row, col = tr, tc
                else:
                    print( 'can not move there')
            else:
                print( 'bad position')
        
        elif action == 'h' or action == 'help':
            action = minimax(board)
            if action is not None:
                x, y = action
                x, y = chr(ord('A') + x), chr(ord('A') + y)
                print('play: {}'.format((x, y)))
            else:
                print('no moves left to play.')

        if row >= 0 and col >= 0:
            action = (row, col)
            board = make_move(board, action, turn)
        
        if terminal(board):
            victor = winner(board)
            if victor != EMPTY:
                print('player {} wins.'.format('O' if victor == X else 'X'))
            else:
                print('draw')
            
            print("new game")
            board = initial_state(board_size)
            continue

#----------------------------------------------------------------------
if __name__ == '__main__':
    gamemain()

File: gomoku_cli.jl
using LinearAlgebra, Random, StaticArrays

is_empty = iszero
const X = 1 # White Player
const O = -1 # Black Player
const EMPTY = 0


"""
Returns starting state of the board.
"""
initial_state(size=15) = repeat([EMPTY], size, size)

"""
Returns player who has the next turn on a board.
"""
player_turn(board) = any(is_empty, board) ? ((sign∘sum)(board) == X ? O : X) : EMPTY

"""
Returns set of all possible actions (i, j) on the board.
"""
actions(board) = findall(is_empty, board)

"""
Returns board after the move.
"""
make_move!(board, action::CartesianIndex{2}, player=player_turn(board)) = (all(i->0<i<=size(board, 1), action.I)) && (board[action] = player)
make_move!(board, action, player=player_turn(board)) = make_move!(board, CartesianIndex(action...), player)

"""
Returns the board that results from making move (i, j) on the board.
"""
result(board, action::CartesianIndex{2}) = (new_board = copy(board); make_move!(new_board, action); return new_board)
result(board, action) = result(board, CartesianIndex(clamp.(action, 1, size(board, 1))...))
result(board, i::T, j::T) where T <: Integer = result(board, CartesianIndex(clamp(i, 1, size(board, 1))), CartesianIndex(clamp(j, 1, size(board, 1))))

"""
Returns the value of board at action.
"""
get(board, action::CartesianIndex{2}) = board[action]
get(board, action) = all(i->0<i<=size(board, 1), action) ? get(board, CartesianIndex(action)) : EMPTY
get(board, i::T, j::T) where T <: Integer = get(board, (i, j))

"""
Returns the winner of the game, if there is one.
"""
function winner(board)
    dirs = ((1, -1), (1, 0), (1, 1), (0, 1))
    # board_len = board.shape[0]
    board_len = size(board, 1)
    
    for i in 1:board_len, j in 1:board_len
        is_empty(board[i, j]) && continue
        id = board[i, j]
        winning_moves = min(5, size(board, 1))
        for d in dirs
            x, y = i, j
            count = 0
            while get(board, (x, y)) == id
                (count > winning_moves || get(board, (x, y)) != id) && break
                x += d[1]
                y += d[2]
                count += 1
            end
            
            # Exactly 5 in a row and no more than 5
            (count == winning_moves && get(board, i+winning_moves*d[1], j+winning_moves*d[2]) != id && get(board, i-d[1], j-d[2]) != id) &&
                (return id)
        end
    end
    
    EMPTY
end

"""
Returns True if game is over, False otherwise.
"""
terminal(board) = winner(board) != EMPTY || EMPTY ∉ board

function display_state(board)
    board_len = size(board, 1)
    
    println(" "*" -"^board_len)
    println("  "*join('A' .+ (0:board_len-1), ' '))

    for row in 1:board_len
        print(('A' + row-1) * ' ')
        println(join(ch == EMPTY ? "∘ " : ch == X ? "X " : "O " for ch ∈ board[row, :]))
    end

    if terminal(board)
        victor = winner(board)
        if victor != EMPTY
            println("""player $(victor == X ? 'X' : 'O') wins.""")
        else
            println("draw")
        end
    else
        turn = player_turn(board) == X ? 'X' : 'O'
        println("Player $turn's turn")
    end
end


"""
# TODO
makes minimax converge faster by:
1: make a 6 x 6 filter kernel and apply that filter to game board and perform these filter operations,
2: rotate the filter(by 90°) 4 times and store those 4 filter configurations.
3: convolve through the board parallelly(not sequentially) using all 4 filters.
4: if any of the 4 filter matches the board subset during convolution, then
5: apply minimax to find winning move(location) on the filtered subset of the board
6: rotate the move/location according to filter configuration."
7: calculate and assign (TODO: some kind of score) to the move to find the optimal(max/min) scoring move from the (board_size x board_size)moves obtained after entire convolvolution.
8: return the optimal(minimum/maximum) move
"""
function ai_move(board)
    nothing
end


struct AI_Knowledge 
    board::Matrix{Int8}
    action::CartesianIndex{2}
    player::Int8
    actions::Vector{CartesianIndex{2}}
    symmetricity::SMatrix
    board_size::UInt8
end

"""
Returns the optimal action for the current player on the board.
"""
function minimax(board; default_return_move=CartesianIndex(-1, -1))
  terminal(board) && (return default_return_move)
  # find which pllayer's turn it is in the board
  turn = player_turn(board)
  # find the playable moves
  possible_moves = actions(board)
  # shuffle for randomness
  shuffle!(possible_moves)
  # an object to map value to optimised moves
  value_to_move = Vector{Tuple{Int8, CartesianIndex{2}}}(undef, length(possible_moves))
  # picks action a in ACTIONS(s) that produces optimal value of MIN/MAX-VALUE(RESULT(s, a))
  for (i, move) in enumerate(possible_moves)
    possible_board = result(board, move)
    score = alphabeta(possible_board, MINIMUM_SCORE, MAXIMUM_SCORE, turn != X)
    value_to_move[i] = (score, move)
  end
  # choose our optimizer function (min or max) based on which player is playing
  optimizer = turn == X ? argmax : argmin
  # return the optimised action from dict.values
  optimal_value, optimal_move = optimizer(x->x[1], value_to_move)
  optimal_move
end

const MINIMUM_SCORE = Int8(-100)
const MAXIMUM_SCORE = Int8(100)
""" min_/max_value function with alpha-beta pruning"""
function alphabeta(board, α, β, maximizingPlayer)
  terminal(board) && (return winner(board))

  score, optimizer = maximizingPlayer ? (MINIMUM_SCORE, max) : (MAXIMUM_SCORE, min)
  
  possible_moves = actions(board)

  for move in possible_moves
    child = result(board, move)
    
    score = optimizer(score, alphabeta(child, α, β, !maximizingPlayer))

    if maximizingPlayer
      α = optimizer(α, score)
    else
      β = optimizer(β, score)
    end
    
    α >= β && break
  end
  
  score
end

#----------------------------------------------------------------------
# main game
#----------------------------------------------------------------------
function gamemain(board_size=15)
    board = initial_state(board_size)

    while true
        row = col = -1
        display_state(board)
        turn = player_turn(board)
        turn_chr = turn == X ? 'X' : 'O'
        print("Enter your move (q/f:Quit/Forfeit): ")
        action = (lowercase∘strip∘readline)()
        
        if length(action) != 2 && action[1] == 'f'
            println("player $turn_chr forfeits.")
            println("player $(turn == X ? 'O' : 'X') wins.")
            
            println("new game")
            board = initial_state(board_size)
            continue
        
        elseif length(action) != 2 && action[1] == 'q'
            println("player $turn_chr quits.")
            break

        elseif length(action) == 2
            tr = action[1] - 'a' + 1
            tc = action[2] - 'a' + 1
            if tr > 0 && tc > 0 && tr <= board_size && tc <= board_size
                if board[tr, tc] == EMPTY
                    row, col = tr, tc
                else
                    println("can not move there")
                end
            else
                println("bad position")
            end
        
        elseif length(action) != 2 && action[1] == 'h'
            ai_action = minimax(board).I
            if !ismissing(ai_action)
                x, y = (ai_action[1] + 'a' - 1), (ai_action[2] + 'a' - 1)
                println("play: $((x, y))")
            else
                println("no moves left to play.")
            end
        end

        (row > 0 && col > 0 && row <= board_size && col <= board_size) && make_move!(board, (row, col), turn)
        
        if terminal(board)
            victor = winner(board)
            victor != EMPTY ? println("player $(victor == X ? 'X' : 'O') wins.") : println("draw")
            
            println("new game")
            board = initial_state(board_size)
            continue
        end
    end
end